File: app/src/main/java/com/example/mnowak/cirriculumvitae/ui/activities/MoreInfoActivity.java
package com.example.mnowak.cirriculumvitae.ui.activities;

import android.databinding.DataBindingUtil;
import android.os.Bundle;
import android.support.v4.app.Fragment;
import android.support.v4.view.ViewPager;
import android.support.v7.app.ActionBar;
import android.support.v7.app.AppCompatActivity;
import android.view.MenuItem;

import com.example.mnowak.cirriculumvitae.R;
import com.example.mnowak.cirriculumvitae.databinding.ActivityMoreInfoBinding;
import com.example.mnowak.cirriculumvitae.models.PersonalInfoViewModel;
import com.example.mnowak.cirriculumvitae.ui.adapters.MoreInfoPagerAdapter;
import com.example.mnowak.cirriculumvitae.ui.fragments.ExperienceFragment;
import com.example.mnowak.cirriculumvitae.ui.fragments.StudiesActivitiesFragment;

import java.util.ArrayList;
import java.util.List;

import butterknife.BindView;
import butterknife.ButterKnife;

public class MoreInfoActivity extends AppCompatActivity {

    public static final String PERSONAL_INFO_EXTRA = "personal_info";

    private static final int EXPERIENCE_PAGE = 0;
    private static final int STUDIES_PAGE = 1;
    private static final int SKILLS_PAGE = 2;
    private static final int[] TITLES_IDS = {R.string.more_info_title_experience, R.string.more_info_title_studies, R.string.more_info_title_skills};
    private PersonalInfoViewModel personalInfoViewModel;

    @BindView(R.id.vpContent)
    ViewPager vpContent;
    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        ActivityMoreInfoBinding binding = DataBindingUtil.setContentView(this, R.layout.activity_more_info);
        retrieveDataFromIntent();
        binding.setViewModel(personalInfoViewModel);
        ButterKnife.bind(this);
        initActionBar();
        prepareViewPager();
    }

    private void retrieveDataFromIntent() {
        Bundle extras = getIntent().getExtras();
        if (extras != null && !extras.isEmpty()) {
            personalInfoViewModel = (PersonalInfoViewModel) extras.getSerializable(PERSONAL_INFO_EXTRA);
        }
    }

    private void initActionBar() {
        ActionBar actionBar = getSupportActionBar();
        if (actionBar != null) {
            actionBar.setDisplayHomeAsUpEnabled(true);
            setTitle(TITLES_IDS[EXPERIENCE_PAGE]);
        }
    }

    @Override
    public boolean onOptionsItemSelected(MenuItem item) {
        switch (item.getItemId()) {
            case android.R.id.home:
                finish();
                return true;
            default:
                return super.onOptionsItemSelected(item);
        }
    }

    private void prepareViewPager() {
        List<Fragment> screens = getUsedPages();
        MoreInfoPagerAdapter screenPagerAdapter = new MoreInfoPagerAdapter(getSupportFragmentManager(), screens);
        vpContent.setAdapter(screenPagerAdapter);
        vpContent.addOnPageChangeListener(onPageChangeListener);
    }

    private ViewPager.OnPageChangeListener onPageChangeListener = new ViewPager.OnPageChangeListener() {
        @Override
        public void onPageScrolled(int position, float positionOffset, int positionOffsetPixels) {
        }

        @Override
        public void onPageSelected(int position) {
            setTitle(TITLES_IDS[position]);
        }

        @Override
        public void onPageScrollStateChanged(int state) {

        }
    };

    public List<Fragment> getUsedPages() {
        List<Fragment> screens = new ArrayList<>();
        screens.add(ExperienceFragment.newInstance(personalInfoViewModel.experience));
        screens.add(StudiesActivitiesFragment.newInstance(personalInfoViewModel.activities));
        return screens;
    }
}

File: app/src/main/java/com/mnowak/curriculumvitae/feature/moreInfo/MoreInfoActivity.kt
package com.mnowak.curriculumvitae.feature.moreInfo

import android.os.Bundle
import android.view.MenuItem
import androidx.databinding.DataBindingUtil
import androidx.lifecycle.Observer
import androidx.lifecycle.ViewModelProviders
import com.mnowak.curriculumvitae.R
import com.mnowak.curriculumvitae.databinding.ActivityMoreInfoBinding
import com.mnowak.curriculumvitae.di.viewModel.ViewModelFactory
import com.mnowak.curriculumvitae.feature.moreInfo.fragment.experience.ExperienceFragment
import com.mnowak.curriculumvitae.feature.moreInfo.fragment.skills.SkillsFragment
import com.mnowak.curriculumvitae.feature.moreInfo.fragment.studiesActivities.StudiesActivitiesFragment
import com.mnowak.curriculumvitae.utils.onPageChanged
import dagger.android.support.DaggerAppCompatActivity
import javax.inject.Inject

class MoreInfoActivity : DaggerAppCompatActivity() {

    @Inject
    lateinit var factory: ViewModelFactory
    lateinit var binding: ActivityMoreInfoBinding

    private val viewModel: MoreInfoViewModel by lazy {
        ViewModelProviders.of(this, factory).get(MoreInfoViewModel::class.java)
    }

    private val screens = listOf(ExperienceFragment.newInstance(),
            StudiesActivitiesFragment.newInstance(),
            SkillsFragment.newInstance())

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        bindView()
        initActionBar()
        prepareViewPager()
        observeTitle()
    }

    private fun bindView() {
        binding = DataBindingUtil.setContentView<ActivityMoreInfoBinding>(this, R.layout.activity_more_info)
        binding.setLifecycleOwner(this)
        binding.viewModel = viewModel
    }

    private fun initActionBar() {
        supportActionBar?.setDisplayHomeAsUpEnabled(true)
    }

    private fun prepareViewPager() {
        val screenPagerAdapter = MoreInfoPagerAdapter(supportFragmentManager, screens)
        binding.vpContent.adapter = screenPagerAdapter
        binding.vpContent.onPageChanged(viewModel::onPageChanged)
    }

    private fun observeTitle() {
        viewModel.activityTitle.observe(this, Observer {
                supportActionBar?.setTitle(it)
        })
    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean =
            when (item.itemId) {
                android.R.id.home -> {
                    finish()
                    true
                }
                else -> super.onOptionsItemSelected(item)
            }
}
